### PR TITLE
test: add verify_at_rows in MockProver for improve test performance

### PR DESCRIPTION
**Pull request by lispc**
the reason for this change is discussed at <https://github.com/appliedzkp/zkevm-circuits/pull/354>

### CODE REVIEW

File: halo2_proofs/src/dev.rs
Comment (CPerezz) [inline halo2_proofs/src/dev.rs:749]: Can we leave `verify` with the regular behavior and leave it co-existing with `verify_at_rows`?

File: halo2_proofs/src/dev.rs
Comment (CPerezz) [inline halo2_proofs/src/dev.rs:761]: Why collecting directly here? 
You can use the row_ids iter and zip or chain it when needed and use `copied` or `cloned` depending on your needs.

**Review by CPerezz — COMMENTED** [review]
LGTM modulo the comments.

File: halo2_proofs/src/dev.rs
Comment (Brechtpd) [inline halo2_proofs/src/dev.rs:962]: It looks like this also disables the verification of lookups to just the passed in `row_ids`, but the EVM circuit does do lookups on rows that don't have `q_step` enabled (currently just byte lookups, but in the future all kinds of lookups). If that's the case (and I could certainly be wrong) then I don't think we can limit the rows like this (at least not with the same `row_ids` as for the gates).

File: halo2_proofs/src/dev.rs
Comment (lispc) [reply halo2_proofs/src/dev.rs:749]: in fact the real logic of current `verify` is identical to old `verify`. By changing it to `self.verify_at_rows(self.usable_rows.clone())` we can avoid keeping 2 copy of hundreds of lines of very similar codes

File: halo2_proofs/src/dev.rs
Comment (lispc) [reply halo2_proofs/src/dev.rs:962]: ok i understand, seems it is better change method signature to `pub fn verify_at_rows<I: Iterator<Item = usize>>(
        &self,
        gate_row_ids: I,
        lookup_input_row_ids: I,
    )`

File: halo2_proofs/src/dev.rs
Comment (lispc) [reply halo2_proofs/src/dev.rs:761]: make sense. i will modify this

File: halo2_proofs/src/dev.rs
Comment (CPerezz) [reply halo2_proofs/src/dev.rs]: hmmm. kk sure. Since it only affects the `MockProver` it should not be an issue.

**Review by CPerezz — APPROVED** [review]
LGTM. I'd like to wait for an approval from @kilic and when done, we can merge! :)

File: halo2_proofs/src/dev.rs
Comment (kilic) [inline halo2_proofs/src/dev.rs:765]: Is it better to range check `row_id` against `usable_rows.end` and `usable_rows.start` rather than `contains`. As it is it could be time consuming to existence query for large `usable_rows`.

File: halo2_proofs/src/dev.rs
Comment (kilic) [inline halo2_proofs/src/dev.rs:841]: I can't really see the difference between `(row + n + at.0) % n` and `(row + at.0) % n`. Probably I'm missing smtg.

File: halo2_proofs/src/dev.rs
Comment (CPerezz) [reply halo2_proofs/src/dev.rs:765]: Indeed, checking than the len of `gate_row_ids` exists in `self.usable_rows` should suffice if as @kilic said (and I agree) the vector contains sequential values.

File: halo2_proofs/src/dev.rs
Comment (lispc) [reply halo2_proofs/src/dev.rs:841]: reason for this is that when row is small and at is negative, `(row + at.0) % n` may be negative? say row is 1, and at is -3, so a negative index is calculated here. Is it correct?

File: halo2_proofs/src/dev.rs
Comment (lispc) [reply halo2_proofs/src/dev.rs:765]: usable_rows here is Range not vector.   `contains` implementation of Range is indeed check low bound and high bound :  https://doc.rust-lang.org/src/core/ops/range.rs.html#810-813

Comment (CPerezz) [reply halo2_proofs/src/dev.rs:765]: But the point is that it checks it for each `row_id`.
When indeed the only `row_ids` that matter are `gate_row_ids[0]` and `gate_row_ids[last_pos]`

File: halo2_proofs/src/dev.rs
Comment (CPerezz) [reply halo2_proofs/src/dev.rs:841]: That makes sense. But it could happen the same before no? Like in the previous status of that, there was the same possibility.

Comment (lispc) [reply halo2_proofs/src/dev.rs:841]: previous codes also includes this logic, but ealier.  <https://github.com/appliedzkp/halo2/pull/38/files#diff-faaf1a95220be5b9b73f935e6659140eac1bd000fbbccb6e102589b4c3ec49c5L802>

File: halo2_proofs/src/dev.rs
Comment (lispc) [reply halo2_proofs/src/dev.rs:765]: check row_id in Range is **much** cheaper than checking gates constraints, and checking only `gate_row_ids[0]` and `gate_row_ids[last_pos]`, we require caller to provided sorted row_ids.    I think the cost/benefit is better for checking every id?

File: halo2_proofs/src/dev.rs
Comment (kilic) [reply halo2_proofs/src/dev.rs:841]: I understand thanks!

File: halo2_proofs/src/dev.rs
Comment (CPerezz) [reply halo2_proofs/src/dev.rs:765]: Ohhh I see now!! Sure. Makes sense now. Also with the improvements in testing time that you announced, we can merge this and check this minor things in the future.

**Pull request merged**
